src/EditorFeatures/CSharpTest/GenerateVariable/GenerateVariableTests.cs

```diff
@@ -7607,6 +7608,284 @@ void Method()
 
     void Goo(ref bool b) { }
     void Goo(int i) { }
+}");
+        }
+
+        [WorkItem(9090, "https://github.com/dotnet/roslyn/issues/9090")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateVariable)]
+        public async Task TestPropertyPatternInIsPattern1()
+        {
+            await TestInRegularAndScriptAsync(
+@"
+class C
+{
+    void M2()
+    {
+        object o = null;
+        if (o is Blah { [|X|]: int i })
```

Comment:
Could you add some tests for:
- `{ [|X|]: }` (incomplete pattern, should produce `X` with type `object`)
- `{ [|X|]: (1, 2) }` (positional pattern, should also produce `object`)
- `{ [|X|]: Type (1, 2) { sub-pattern } }` (should produce `X` with type `Type`)

Reply:
> { [|X|]: (1, 2) } (positional pattern, should also produce object)

Why not produce an ```(int,int)```?

Reply:
@jcouv Why would we produce `X` with type `Type`? If we do that, then there's no point having a type inside that pattern if the type already known. Generally we use the type in a pattern to check for a more specific type and it doesn't match the type of the value we're matching against.

Reply:
> Why would we produce X with type Type? If we do that, then there's no point having a type inside that pattern if the type already known. 

yes there is.  It ensures the type is non-null.

Reply:
I don't believe we should infer the type of the property based on the pattern we're matching it against. That would defeat the point of doing the matching in the first place.

Reply:
I have listed several reasons why this is appropriate.  Both for just deconstructing into a a fresh variable, when the user likes to be explicit about types, and also for ensuring non-nullness.

Reply:
> yes there is. It ensures the type is non-null.

Yes null checking is usually not the point of using a pattern and definitely not the major use case. I believe that in most cases when users would specify a type, they want a type check.

Reply:
I think the type should be inferred from the property itself down when generating a named constant but not the other way around. I just don't see how that's useful.

Reply:
> Yes but null checking is usually not the point of using pattern matching and definitely not the major use case. I believe that in most cases when users would specify a type, they want a type check.

I can think of no evidence to support that position.  Regardless, this is a perfectly reasonable interpretation of of the code and matches what hte user will want some of the time.  I'm ok with that here.

Reply:
More specifically: the "i want a fresh deconstructed variable where i specify the type" seems utterly valid and completely in line with how many will use the language.  It's precisely for someone who likes pattern matching but still wants explicit types.  I see no reason to not infer that they would like the *same* type when doing generation.

Reply:
Ok... my expectation when generating x in `x is int i` is definitely not for `x` to be an int.

Reply:
> I think the type should be inferred from the property itself down when generating a named constant but not the other way around. I just don't see how that's useful.

I have given two concrete examples of how i think it would be useful to some subsets of our customer base.  That is sufficient to me.  I see no reason to think that that pattern matching will not *commonly* be used for this case.  i.e. someone simply wants to deconstruct, and they just like explicit types.  As such, they will do ```X: type x, Y: type y, ...``` as the *norm* for how they deconstruct.  Inferring thusly what the types of X and Y are seems very sensible to me. 

In the case where we're wrong, the user fixes up the type.  That's the same with the type-inferred *already* today.  THis is 100% a heuristic-based approach to *guess* at what we think is the most reasonable thing to infer.  It's often wrong because the user intends something different.  That's ok.

Reply:
Ok, thanks.